### PR TITLE
add scale flag for compose up

### PR DESCRIPTION
**Pull request by Junnplus**
Signed-off-by: ye.sijun <junnplus@gmail.com>

Add scale flag for compose up.

```
./nerdctl compose up -d --scale wordpress=3
INFO[0000] Creating network nerdctl_default
INFO[0000] Ensuring image mariadb:10.5
INFO[0001] Ensuring image wordpress:5.7
INFO[0001] Creating container nerdctl_wordpress_3
INFO[0001] Creating container nerdctl_wordpress_2
INFO[0001] Creating container nerdctl_wordpress_1
INFO[0001] Creating container nerdctl_db_1
```

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Thanks, could you also update `README.md`?
Adding tests would be also appreciated.

File: pkg/composer/logs.go [hunk 1]
```diff
@@ -64,13 +59,18 @@ func (c *Composer) logs(ctx context.Context, containers map[string]serviceparser
...
+			return err
+		}
+		name := info.Labels[labels.Name]
+		logTag := strings.TrimPrefix(name, c.project.Name+"_")
 		if l := len(logTag); l > logTagMaxLen {
 			logTagMaxLen = l
 		}
-		containerStates[id] = containerState{
-			name:   container.Name,
+		containerStates[name] = containerState{
```
Comment (AkihiroSuda) [inline pkg/composer/logs.go:72]: The key should be container ID. Names are not guaranteed to be unique.

File: pkg/composer/up.go [hunk 2]
```diff
@@ -36,6 +36,7 @@ type UpOptions struct {
 	ForceBuild  bool
 	IPFS        bool
 	QuietPull   bool
+	Scale       map[string]uint64
```
Comment (AkihiroSuda) [inline pkg/composer/up.go:39]: Please explain the key in the comment line

File: pkg/composer/composer.go [hunk 3]
```diff
@@ -50,7 +51,7 @@ type Options struct {
 	DebugPrintFull bool // full debug print, may leak secret env var to logs
 }
 
-func New(o Options) (*Composer, error) {
+func New(o Options, client *containerd.Client) (*Composer, error) {
```
Comment (AkihiroSuda) [inline pkg/composer/composer.go:54]: Can we put `client` to `Options` struct?

File: pkg/composer/container.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/nerdctl/pkg/labels"
+	"github.com/sirupsen/logrus"
+)
+
+func (c *Composer) GetContainers(ctx context.Context, services ...string) ([]containerd.Container, error) {
```
Comment (AkihiroSuda) [inline pkg/composer/container.go:28]: `GetContainers` -> `Containers`

File: pkg/composer/down.go [hunk 5]
```diff
@@ -31,8 +34,17 @@ type DownOptions struct {
 }
 
 func (c *Composer) Down(ctx context.Context, downOptions DownOptions) error {
-	for _, svc := range c.project.Services {
-		if err := c.downService(ctx, svc, downOptions.RemoveVolumes); err != nil {
```
Comment (AkihiroSuda) [inline pkg/composer/down.go:35]: `downService()` is no longer used? Can we remove `downService()`?

Comment (Junnplus) [reply pkg/composer/down.go:35]: Sure.

File: pkg/composer/composer.go (hunk 3 above)
Comment (Junnplus) [reply pkg/composer/composer.go:54]: I prefer to separate them, they don’t seem to be affiliation.

File: pkg/composer/up_service.go [hunk 6]
```diff
@@ -45,11 +45,14 @@ func (c *Composer) upServices(ctx context.Context, parsedServices []*servicepars
...
 	var (
 		containers   = make(map[string]serviceparser.Container) // key: container ID
+		services     = []string{}
 		containersMu sync.Mutex
 		runEG        errgroup.Group
 	)
 	for _, ps := range parsedServices {
+		services = append(services, ps.Unparsed.Name)
 		for _, container := range ps.Containers {
+			ps := ps
```
Comment (AkihiroSuda) [inline pkg/composer/up_service.go:55]: Shouldn't this be right after `for _, ps := range parsedServices {` line?

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**